### PR TITLE
Process Manager and Seat Assignments Aggregate

### PR DESCRIPTION
**Pull request by AlexanderLitus**
Is based on PR #11.

### CODE REVIEW

**Comment by AlexanderLitus** [comment]
@alexander-yevsyukov, PTAL (proc manager is covered 50% now).

File: registration/src/test/java/org/spine3/samples/lobby/registration/procman/RegistrationProcessManagerRepositoryShould.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import static org.spine3.samples.lobby.common.util.IdFactory.newConferenceId;
+import static org.spine3.samples.lobby.common.util.IdFactory.newOrderId;
+import static org.spine3.samples.lobby.registration.procman.RegistrationProcess.State.PAYMENT_RECEIVED;
+import static org.spine3.samples.lobby.registration.testdata.TestDataFactory.newBoundedContext;
+
+/**
+ * @author Alexander Litus
+ */
+@SuppressWarnings("InstanceMethodNamingConvention")
+public class RegistrationProcessManagerRepositoryShould {
```
Comment (alexander-yevsyukov) [inline registration/src/test/java/org/spine3/samples/lobby/registration/procman/RegistrationProcessManagerRepositoryShould.java]: Do we expect the end user write tests like this?

File: registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+            super(id);
+        }
+
+        @Override
+        @SuppressWarnings("RefusedBequest") // is overridden to do not throw an exception
+        protected RegistrationProcess getDefaultState() {
+            return RegistrationProcess.getDefaultInstance();
+        }
+
+        // Is overridden to make accessible in tests.
```
Comment (alexander-yevsyukov) [inline registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java]: `@VisibleForTesting` annotation also applies here.

File: registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+                .toBuilder()
+                .setOrderId(ORDER_ID)
+                .setConferenceId(CONFERENCE_ID)
+                .setReservationAutoExpiration(minutesAhead(15))
+                .setProcessState(processState)
+                .setIsCompleted(isCompleted);
+        return builder.build();
+    }
+
+    /*package*/ static Timestamp reservationExpirationTimeBeforeNow() {
```
Comment (alexander-yevsyukov) [inline registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java]: Do we have spaces around `package` in our coding convention? This form is more compact and looks better. If it's not standard we need to update our standard and main code to reflect it. Please take care of it.

File: registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+            @Override
+            protected void send(Message commandMessage) {
+                commandsSent.add(commandMessage);
+                super.send(commandMessage);
+            }
+        }
+    }
+
+    /*package*/ static class Event {
```
Comment (alexander-yevsyukov) [inline registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java]: What do we have this class for? Its name clashes with the class from the main API. Is it to be able to write something like: `Given.Event.orderPlaced()`? If so, please document this.

File: registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+        /*package*/ static OrderConfirmed orderConfirmed() {
+            final OrderConfirmed.Builder builder = OrderConfirmed.newBuilder()
+                    .setOrderId(ORDER_ID)
+                    .addAllSeat(SEATS);
+            return builder.build();
+        }
+    }
+
+    /*package*/ static class Command {
```
Comment (alexander-yevsyukov) [inline registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java]: Is this class named such to be able to write `Given.Command.expireRegistrationProcess()`? If so, please document this.

File: registration/src/test/java/org/spine3/samples/lobby/registration/procman/RegistrationProcessManagerRepositoryShould.java (hunk 1 above)
Comment (AlexanderLitus) [inline registration/src/test/java/org/spine3/samples/lobby/registration/procman/RegistrationProcessManagerRepositoryShould.java]: Why not? I've found several bugs due to these tests.

File: registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java (hunk 4 above)
Comment (AlexanderLitus) [inline registration/src/test/java/org/spine3/samples/lobby/registration/procman/Given.java]: Yes, but not only for better readability of its usages - to structurize the code also.

File: registration/src/main/java/org/spine3/samples/lobby/registration/conference/ConferenceProjection.java [hunk 6]
```diff
@@ -161,7 +161,7 @@ private void addOrRemoveSeats(int newQuantity, int oldQuantity, SeatTypeId id) {
         }
     }
 
-    // TODO:2016-02-23:alexander.litus: why is it here and not in PM?!
+    // TODO:2016-02-23:alexander.litus: consider moving this method to Process Manager
```
Comment (alexander-yevsyukov) [inline registration/src/main/java/org/spine3/samples/lobby/registration/conference/ConferenceProjection.java]: Can we decide on this matter in this PR?

File: registration/src/main/java/org/spine3/samples/lobby/registration/conference/ConferenceProjection.java [hunk 7]
```diff
@@ -171,6 +171,7 @@ private void sendAddSeatsRequest(SeatTypeId seatTypeId, int quantity) {
         boundedContext.process(command);
     }
 
+    // TODO:2016-02-23:alexander.litus: consider moving this method to Process Manager
```
Comment (alexander-yevsyukov) [inline registration/src/main/java/org/spine3/samples/lobby/registration/conference/ConferenceProjection.java]: Can we decide on this matter in this PR?

File: registration/src/main/java/org/spine3/samples/lobby/registration/order/OrderAggregate.java [hunk 8]
```diff
@@ -203,6 +197,27 @@ private void apply(OrderReservationCompleted event) {
     }
 
     @Apply
+    private void apply(OrderTotalsCalculated event) {
+        final Order.Builder state = getState().toBuilder();
+        state.setPrice(event.getTotal());
+        incrementState(state.build());
```
Comment (alexander-yevsyukov) [inline registration/src/main/java/org/spine3/samples/lobby/registration/order/OrderAggregate.java]: Aggregates already do not need to call `incrementState()`. Can you migrate this code to latest state of `core-java`?

File: registration/src/test/java/org/spine3/samples/lobby/registration/conference/ConferenceProjectionRepositoryShould.java [hunk 9]
```diff
@@ -71,7 +72,7 @@ private TestConferenceProjection(ConferenceId id) {
             super(id);
         }
 
-        // Is overridden to make it accessible in tests.
+        @VisibleForTesting
```
Comment (alexander-yevsyukov) [inline registration/src/test/java/org/spine3/samples/lobby/registration/conference/ConferenceProjectionRepositoryShould.java]: Can we have it package access for tests? Are these tests still required assuming that calling `incrementState()` is no longer needed?

File: registration/src/test/java/org/spine3/samples/lobby/registration/order/Given.java [hunk 10]
```diff
@@ -197,36 +206,56 @@ private Command() {
...
+        /* package */ static OrderExpired orderExpired() {
+            final OrderExpired.Builder builder = OrderExpired.newBuilder()
+                    .setOrderId(ORDER_ID);
+            return builder.build();
+        }
+
+        /* package */ static OrderRegistrantAssigned orderRegistrantAssigned() {
+            final OrderRegistrantAssigned.Builder builder = OrderRegistrantAssigned.newBuilder()
+                    .setOrderId(ORDER_ID)
+                    .setPersonalInfo(newPersonalInfo("Albert", "Einstein", "ein@stein.com"));
```
Comment (alexander-yevsyukov) [inline registration/src/test/java/org/spine3/samples/lobby/registration/order/Given.java]: Put a email address with a domain of the establishment where Albert last worked.

File: registration/src/test/java/org/spine3/samples/lobby/registration/seat/assignment/TestSeatAssignmentsAggregate.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+import org.spine3.samples.lobby.registration.contracts.SeatAssignmentsId;
+import org.spine3.samples.lobby.registration.contracts.SeatUnassigned;
+
+import javax.annotation.Nonnull;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import static com.google.common.base.Throwables.propagate;
+
+/**
```
Comment (alexander-yevsyukov) [inline registration/src/test/java/org/spine3/samples/lobby/registration/seat/assignment/TestSeatAssignmentsAggregate.java]: Please document this class. Its purpose is not clear.

File: registration/src/test/java/org/spine3/samples/lobby/registration/seat/assignment/TestSeatAssignmentsAggregate.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+ * @author Alexander Litus
+ */
+/* package */  class TestSeatAssignmentsAggregate extends SeatAssignmentsAggregate {
+
+    /* package */ TestSeatAssignmentsAggregate(SeatAssignmentsId id) {
+        super(id);
+    }
+
+    // Is overridden to do not throw exceptions while retrieving the default state via reflection.
```
Comment (alexander-yevsyukov) [inline registration/src/test/java/org/spine3/samples/lobby/registration/seat/assignment/TestSeatAssignmentsAggregate.java]: Perhaps we need to fix this matter at the higher level?

**Comment by AlexanderLitus** [comment]
PTAL.

File: registration/src/test/java/org/spine3/samples/lobby/registration/seat/availability/testcase/ExistAvailableSeatsAndPendingReservations.java [hunk 13]
```diff
@@ -53,13 +53,13 @@
...
-    public TestSeatsAvailabilityAggregate givenAggregate() {
-        final TestSeatsAvailabilityAggregate aggregate = super.givenAggregate();
+    public SeatsAvailabilityAggregate givenAggregate() {
+        final SeatsAvailabilityAggregate aggregate = super.givenAggregate();
         final SeatsAvailability state = aggregate.getState().toBuilder()
                 .addAllAvailableSeat(AVAILABLE_SEATS)
                 .putAllPendingReservations(PENDING_RESERVATIONS)
                 .build();
-        aggregate.incrementState(state);
+        aggregate.testIncrementState(state);
```
Comment (alexander-yevsyukov) [inline registration/src/test/java/org/spine3/samples/lobby/registration/seat/availability/testcase/ExistAvailableSeatsAndPendingReservations.java]: Please rename 
- `testIncrementState()` to `incrementStateForTest()`;
- `testDispatch()` to `dispatchForTest()`.

This way they 1) would sort better with the production counterparts; 2) won't be confused with testing methods.

Comment (alexander-yevsyukov) [inline registration/src/test/java/org/spine3/samples/lobby/registration/seat/availability/testcase/ExistAvailableSeatsAndPendingReservations.java]: Please also add this convention to our coding guidelines.

**Comment by AlexanderLitus** [comment]
Done, can I merge it now?

**Comment by alexander-yevsyukov** [comment]
LGTM

**Pull request merged**